### ISSUE TITLE
Predictions seem wrong a scikit pipeline with OneHotEncoder

### ISSUE THREAD

**Issue by ivannp** [onnx/sklearn-onnx#321]
Have a simple model with two predictors:
* int64, plugged into a OneHot transform
* float32/float64, plugged into a RobustScaler

The model is created fine, but all predictions come out as 0s. None of them comes out as 0 in the original scikit model.

Removing the OneHot transform solves the issue - predictions are similar to the scikit model.

**Comment by xadupre** [comment]
I'll close the issue as the PR is now merged into master branch.

**Comment by victornoel** [comment]
@xadupre does the fix solves also the problem if the inputs of the `OneHotEncoder` are strings?

I'm facing a similar situation with one `OneHotEncoder` in front of a `XGBRegressor`: the output is of course not 0s (because of the regressor), but it is always the same value given.

**Comment by victornoel** [comment]
@xadupre some clarification:
- I'm comparing the predictions via onnxruntime and sklearn/xgboost itself and the there are huge differences when using the `OneHotEncoder`
- I tried with the latest master of `sklearn-onnx` and the problem persists
- if I replace the `OneHotEncode` with an `OrdinalEncoder`, the predictions are equals between sklearn and onnxruntime

**Comment by victornoel** [comment]
@xadupre have you been able to reproduce?

Just to be sure it wasn't related to some float32/float64 strange behaviour, I try to train my model using float32 and I still see huge differences.

For the record, the predictions from sklrean/xgb range from 0 to 1700 and those from onnxruntime range from 0 to 323, with diffs as high as 1400!

I'm not sure what is the best way to reproduce it in a small example, because I'm relying on proprietary data.

**Comment by victornoel** [comment]
For the record here is the code I'm using for converting the model:
```py
train_df, test_df, train_labels, test_labels = train_test_split(
    full_df, full_labels, test_size=.2, random_state=None
)

col_transformer = _column_tranformer_fitted_from_df(full_df)

search = RandomizedSearchCV(
    XGBRegressor(verbose=0, objective='reg:squarederror'),
    param_distributions={
        "colsample_bytree": sp_uniform(),
        "gamma": sp_uniform(.1, 1),
        'learning_rate': sp_uniform(.1, .6),
        'max_depth': sp_randint(10, 30),
        'min_child_weight': sp_uniform(0, 3),
        'n_estimators': range(10, 75),  # nb de pas d'apprentissange pour chaque cas
    },
    cv=5, n_iter=100, n_jobs=multiprocessing.cpu_count()
)
search.fit(self._col_transformer.transform(train_df), train_labels)
regressor = XGBRegressor(verbose=0, objective='reg:squarederror', **search.best_params_)
regressor.fit(
    col_transformer.transform(train_df),
    train_labels
)
model = Pipeline(
    steps=[('preprocessor', col_transformer),
           ('regressor', regressor)]
)

update_registered_converter(
    XGBRegressor, 'XGBRegressor',
    calculate_linear_regressor_output_shapes,
    convert_xgboost
)
onnx = convert_sklearn(
    model,
    initial_types=_convert_dataframe_schema(self._for_eval_df)
)


def _column_tranformer_fitted_from_df(data: pd.DataFrame) -> ColumnTransformer:
    def transformer_for_column(column: pd.Series):
        if column.dtype in ['float64', 'float32']:
            return MinMaxScaler()
        if column.dtype in ['O']:
            return OneHotEncoder()
        raise ValueError(f'Unexpected column dtype for {column.name}:{column.dtype}')

    return ColumnTransformer(
        [(col, transformer_for_column(data[col]), [col]) for col in data.columns],
        remainder='drop'
    ).fit(data)


def _convert_dataframe_schema(data: pd.DataFrame) -> List[Tuple[str, onnxtypes.DataType]]:
    def type_for_column(column: pd.Series):
        if column.dtype in ['float64', 'float32']:
            # onnx does not really support float64 (DoubleTensorType does not work with TreeEnsembleRegressor)
            return onnxtypes.FloatTensorType([None, 1])
        if column.dtype in ['O']:
            return onnxtypes.StringTensorType([None, 1])
        raise ValueError(f'Unexpected column dtype for {column.name}:{column.dtype}')

    return [(col, type_for_column(data[col])) for col in data.columns]
```

**Comment by victornoel** [comment]
@prabhat00155 @xadupre here is a repro: https://gist.github.com/victornoel/06d6231f6276719ddba53cb381dfd468

I casted an `int` column to `str` because in my original dataset, we have categories made of numbers so I thought this could be related…

As we can see, the differences are huge: the predictions should be between 0 and 1 and the differences can be as big as 1 in my tests!
```
>>> import test
>>> test.test()
[0.9019426  0.91082716 0.91868186 0.94926846 1.0898147 ]
min(Y)-max(Y): 0 1
```

**Comment by prabhat00155** [comment]
Looks like the there is some issue with XGBRegressor converter, not OHE. If you replace XGBRegressor with MLPRegressor, it seems to work fine. Here is the output of test() with MLPRegressor().
```
[2.72400598e-07 2.87338889e-07 2.95446651e-07 3.51248081e-07
 4.55852575e-07]
min(Y)-max(Y): 0 1
```

**Comment by victornoel** [comment]
I didn't realise that the converter for XGBRegressor wasn't here earlier, I've opened an issue at onnxmltools: https://github.com/onnx/onnxmltools/issues/371

**Comment by victornoel** [comment]
For the record, I tested latest git version of sklearn-onnx, onnxconverter-common and onnxmltools and the problem is still there…

Any way to get some feedbacks on this maybe?

**Comment by victornoel** [comment]
I suppose this issue can now be closed, according to onnx/onnxmltools#371, the problem is caused by onnxruntime not supporting sparse matrix. Using `sparse=False` in the OneHotEncoder solves the problem.
